Separate bright and bold

LinusU opened the pull request:
Previously all bright colors were always rendered bold, and all bold text was always rendered using the bright colors. This patch separates that so that bold and bright can be controlled individually.

This is probably going to affect people who relied on this previously, so I'm working on an option that re-enables the old behavior. I will push that as a separate commit to this same branch.

Feedback appreciated! 🙌 

Fixes zeit/hyper#2836

LinusU commented:
Thanks for the feedback.

One question, if we have a `drawBoldAsBright` option, should that also draw bright text bold? That seems confusing but that's how it worked before, since bright and bold were essentially the same thing.

We could also make it two options: `drawBoldAsBright` and `drawBrightAsBold`, but then `as` might be confusing as it's rather both. Maybe `drawBoldTextBright` and `drawBrightTextBold`? Both defaulting to true since that was how it was before.

Personally leaning most towards `drawBoldTextBright` and `drawBrightTextBold`...

mofux commented:
I don't think that drawing bright text as bold would make sense. Most terminals offer two options:
- enable / disable drawing bold text as bright (the text may still be bold)
- enable / disable drawing bold text as bold (the text may still be bright)

So we might end up with something like `drawBoldTextBright` and `drawBoldTextBold` that can be combined to get the desired flavour 😁

LinusU commented:
> I don't think that drawing bright text as bold would make sense.

I agree, but as far as I know, that is the current behaviour of xterm.js. So if we want to keep backwards compatibility, I think we need an option for that 🤔

mofux commented:
Hmmm I haven't tested it, but I don't think that the current default behaviour is to draw bright text as bold (I've also not found any code that would point towards that). Imo we should not support an option that draws bright text as bold as it is confusing -- just as you noted. @xtermjs/core Any thoughts?

mofux commented:
>Implement an additional option drawBoldTextInBrightColors that will draw bold characters in bright colors and will default to false

@parisk The default should be `true`, as drawing bold text in bright colors is the default behaviour at the moment.

parisk commented:
Nah, you are right 😅. Fixed it.

LinusU commented:
Hahahhaaha 🤦‍♂️ 🤦‍♂️ 🤦‍♂️ 

```js
const isAscii = code < 256;
// A color is basic if it is one of the standard normal or bold weight
// colors of the characters held in the char atlas. Note that this excludes
// the normal weight _light_ color characters.
const isBasicColor = (colorIndex > 1 && fg < 16) && (fg < 8 || bold);
const isDefaultColor = fg >= 256;
const isDefaultBackground = bg >= 256;
if (this._charAtlas && isAscii && (isBasicColor || isDefaultColor) && isDefaultBackground) {
    // draw from atlas
} else {
    // draw using ctx
}
```

LinusU commented:
Rebased on master, and added two new commits.

The first one fixed so that I'm actually utilizing the char atlas for all 4-bit ANSI colors, bold and non-bold. The second one adds the `drawBoldTextInBrightColors` option which defaults to true.

2f1ec3b, and fd185a5 with `drawBoldTextInBrightColors` set to `false`:

<img width="347" alt="screen shot 2018-04-20 at 10 31 58" src="https://user-images.githubusercontent.com/189580/39044069-fe9ff3f4-4486-11e8-875d-c6517a42332c.png">

fd185a5, which is same as current master:

<img width="346" alt="screen shot 2018-04-20 at 10 33 44" src="https://user-images.githubusercontent.com/189580/39044092-15f6f7dc-4487-11e8-9d72-14244cb927a9.png">

From my side, this is ready for merging now 🚀 let me know if anything else should be fixed 🙌

saul commented on typings/xterm.d.ts:
```diff
@@ -54,6 +54,11 @@ declare module 'xterm' {
     disableStdin?: boolean;
 
+    /**
+     * Wheter to draw bold text in bright colors. The default is true.
```
s/Wheter/Whether/

LinusU replied:
Fixed 👌

LinusU commented:
Rebased to avoid conflicts with #1422 

ping @Tyriar @bgw @parisk, would love to get this merged ❤️

bgw commented:
I added this small commit to fix a minor bug: https://github.com/xtermjs/xterm.js/pull/1391/commits/9893338020cc66efb8180f8f4df6852a0880bfae

Otherwise, LGTM. I manually tested it, and it seemed to work well for me. I'll merge once Travis finishes. Thanks for your contribution, sorry it took me so long to merge it.

bgw commented:
Travis is failing, but the issue is unrelated. It looks like a recent change to one of our dependencies broke some types. I was able to reproduce them locally in master by wiping and re-installing my node_modules.
